Feature/query

rynzr opened the pull request:
Beberapa tambahan dalam PR ini:
- Penambahan fitur basic query (select, insert, update, delete, order by, limit, offset, group by tapi belum ada fitur untuk aggregate).
- Penambahan struct Model dan Field untuk menyimpan properti dari struct, properti bisa digunakan untuk generate query atau struct baru.
- Penambahan migrasi dan seeder untuk keperluan testing
- Penambahan walk function untuk traverse struct Expression atau LogicalExpression dan meng-convert struct tersebut menjadi string
- Penambahan Docker dan Makefile untuk environment testing

rynzr committed:
Implement LogicalExpression walk

Add implementation to walk function inside expression.go

rynzr committed:
Basic Query Implementation

- Add basic query function (select, insert, update, delete, order by, limit, offset, group by without aggregate)
- Create model and field struct to store properties for model to be scanned
- Implement walk function to create where query and arguments
- Add table migration and seeder for testing
- Add docker and makefile for integration testing environment

evo3cx commented:
file changes nya banyak banget hahaha..

evo3cx commented:
Pada file `README.md` tambahkan beberapa bagian ini.
*. Description
*. Requirement ( dependency yang dibutuhkan e.g "harus intall go version 1.10 atau keatas )
*. How to use this library.

evo3cx commented:
module managment yang digunakan apa? jika `go mod` maka file `go.mod` dan `go.sum` nya juga di commit dan push.

reza-putra commented on main_test.go:
```diff
@@ -0,0 +1,86 @@
+
+func TestMain(m *testing.M) {
+	var err error
+	db, err = fury.Connect(&fury.Configuration{
```
gunakan environment variable untuk configuration, dapat menggunakan library ini https://github.com/joho/godotenv.

kenapa harus load configuration dari environment variable? dapat baca disini https://12factor.net/config

rynzr replied:
Oke deh. nanti saya tambahin biar pake environment variable

rynzr replied:
Sudah ditambah yaa

reza-putra commented on query.go:
```diff
@@ -1,73 +1,536 @@
+		return errors.New("Columns and argument length not match")
+	}
+	query.args = append(query.args, args...)
+
```
tidak perlu enter yah, jadi seperti ini.

```
query.args = append(query.args, args...)
if len(cols) < 1 || len(args) < 1 {
```

rynzr replied:
Ini juga sudah dihapus enter-nya

reza-putra commented on query.go:
```diff
@@ -1,73 +1,536 @@
+
+// NextOrCreateModel shift modelPtr to next model or create new and append it to slice
+//	When model is struct or pointer to struct, then this method should return nil, modelPtr not shifted, and not creating new instance
+func (q *Query) nextOrCreateModel() (*model.Model, error) {
```
ini maksudnya `shifted` itu apa yah? saya kurang paham :D

rynzr replied:
Maksudnya ini seperti linked list tapi di implementasikan di slice. Kalau `nextModel()` akan ke element selanjutnya di slice. Kalau yang `nextOrCreate()` akan ke element selanjutnya kalau masih ada, kalau ngga akan nambah baru. Ini dipakai waktu select, jadi kalau misalkan kita select `Account` return 5 row, kita tinggal pass address dari slice `Account` yang kosong, nanti return-nya akan langsung di-scan dan ditambahkan ke slice tersebut.

reza-putra commented on query.go:
```diff
@@ -1,73 +1,536 @@
-	tableName       interface{}
+	SQL             string
+	tableName       string
+	models          []*model.Model
```
ini dalam satu query dapat memiliki banyak models yah? dan model adalah representasi entity pada table? 

kenapa dibuat satu query dapat melakukan banyak query ke beberapa table? kenapa tidak dibuat untuk 1 table? bukannya dengan begitu lebih mudah?

rynzr replied:
Itu satu query untuk satu model sebenernya. Maksudnya dibuat begitu untuk nampung select dengan return lebih dari satu sama kalau mau insert, update, atau delete tanpa `for` lagi

reza-putra replied:
bukannya dengan begini handle error nya akan jadi lebih susah yah?

rynzr replied:
Untuk select dengan return banyak atau satu sama insert, update, delete dengan satu record itu masih mudah karena kalau ada error kan langsung return. Tapi kalau insert, update, delete kalau banyak record karena belum ada transaction-nya jadinya dia allow partial success. Tapi jadi bisa ditambahin opsi kalau emang mau allow partial success atau kalau maunya success semua, nanti harus ditambahin transaction, kalau failed rollback. Paling agak susah di situnya aja, atau ada yang susah lagi?

reza-putra replied:
handle multiple type `Model`  yang berbeda dimana setiap `Model` tersebut merepresentasikan table yang berbeda di DB

rynzr replied:
Ini ngga bisa meng-handle multiple type model, maksudnya ini kalau misalkan ada slice dari model yang sama. Jadi table-nya tetep beda, slice-nya itu merepresentasikan row untuk table tersebut

reza-putra replied:
ini perlu ditambahkan documentationnya, bisa ditambahkan di README, karena di README.md yang sekarang hanya tertulis ini
```
// Insert also support pointer to slice of pointer to struct
```

rynzr replied:
Oke comment-nya nanti ditambahin

reza-putra commented on model/model_test.go:
```diff
@@ -0,0 +1,232 @@
+
+type Account3 struct {
+	UserID    int `fury:"primary_key,auto_increment"`
+	Counter   *int
```
untuk field yang tidak ditambahkan  tag `furry`.  maka column name nya di table menjadi apa?

rynzr replied:
Ini masih belum ada fitur create migration-nya, jadi masih anggapan field di struct sama dengan di table. Tag itu cuma buat tandain kalau dia PK atau auto increment di database. Itu nanti digunain buat generate query condition-nya, atau kalau yang auto increment biar ngga di-include di insert query. Yang sekarang support juga masih dua tag itu aja.

reza-putra commented on model/model.go:
```diff
@@ -0,0 +1,166 @@
+
+// GetScanPtrByColumnNames return scanner pointers ordered as specifed in the input slice.
+func (m *Model) GetScanPtrByColumnNames(columns []string) []interface{} {
+	pointers := []interface{}{}
```
untuk menghindari allocation yang tidak perlu deklarasi slice jangan via literal. 

```
a nil slice is functionally equivalent to a zero-length slice, even though it points to nothing. It has length zero and can be appended to, with allocation.
```
var pointers []interface{}

rynzr replied:
Oke, ini nanti diganti juga

rynzr replied:
Ini juga sudah diganti

reza-putra commented on model/model_test.go:
```diff
@@ -0,0 +1,232 @@
+
+		field1 := f.Field(0)
+		field2 := f.Field(1)
+		tc.want = []interface{}{field1.Addr().Interface(), field2.Addr().Interface()}
```
jika seperti ini bukannya address pointer yang dicompare?

rynzr replied:
Method yang di-test di sini kan untuk dapetin address dari field yang mau di-scan. Kalau mau test method ini kita bandingin address dari struct yang dijadiin model kan?

reza-putra commented on db.go:
```diff
@@ -3,30 +3,191 @@ package fury
+
+// Delete query method
+func (db *DB) Delete(model interface{}, opts ...QueryOption) error {
+	newDB, err := db.clone(model)
```
setiap kali operation `db.clone` setelah selesai melakukan database, kenapa connection databasenya tidak di close?

rynzr replied:
`db.clone` return copy dari struct `DB` dengan `ConnectionPooler` dan `Configuration` yang sama, tapi setiap clone dia bakalan create `Query` baru untuk generate query baru. Nah karena `ConnectionPooler` sama jadi cuma perlu di-close saat terakhir program-nya jalan. Waktu create connection biasa atau request biasa ngga perlu close `ConnectionPooler`-nya

reza-putra replied:
i see, karena menggunakan pointer of `ConnectionPoller` yah. ini gmn jika ada dua user melakukan operation secara bersamaan akan terjadi `data race`. coba baca ini 

https://notes.shichao.io/gopl/ch9/

rynzr replied:
Kalau di bagian data race di database-nya masih, karena ini memang masih implementasi basic-nya belum menambahkan transaction. Tapi harusnya untuk di code-nya sudah kan ya? karena ini clone dulu dan `DB` yang digunakan dari `database/sql` itu concurrent-safe. Tapi kalau di code untuk query-nya kalau misalkan dibuat parallel kemungkinan bisa data race karena ada banyak slice dan map yang belum diberi lock untuk aksesnya

reza-putra replied:
ini kita pisahkan test pada database dll. untuk library nya sendiri hal ini masih dapat terjadi data race.  jika function clone ini tidak ambil pointer of struct tapi copy value nya saja bagaiamana menurutmu?  data-race ini memang sulit untuk ditemukan. jika menurut kamu bagian ini cukup sulit diskip saja dan fix bagian lain gpp.

nanti lanjut ke task onboarding selanjutnya.

rynzr replied:
Sebenernya cuma masih bingung data race-nya ada di bagian mana. Kalau misalkan copy value dari `ConnectionPooler` itu nanti terhitungnya dua connection pool ke database? Dan kalau begitu, apakah proses untuk copy akan open connection lagi ke database?

Untuk bagian ini sepertinya mau coba untuk di-fix dulu, gpp biar sekalian belajar. Untuk task selanjutnya boleh sekalian kerjain aja? Yang penting nanti waktu code bagian repo-nya terakhir.

reza-putra replied:
concern saya karena db clone itu share pointer of connection pooler. jadi jika ada dua user yang melakukan operation secara bersamaan itu akan terjadi data race. 

contohnya mungkin seperti ini. di db semuanya berjalan dengan baik tapi pada library data yang dikembalikan tidak benar, data race mungkin terjadi di 1 dari 1000 operation tapi ini akan untuk me-tracenya ketika terjadi.

![Screenshot from 2019-06-30 00-18-02](https://user-images.githubusercontent.com/7883991/60387399-9ed0ca00-9acc-11e9-91b1-e3b17483d11a.png)

untuk memulai mengerti apa itu data-race dan bagaimana mengatasinya bisa mulai membaca ini:
- https://golang.org/doc/articles/race_detector.html
- https://medium.com/@val_deleplace/does-the-race-detector-catch-all-data-races-1afed51d57fb

kita berlajar bareng saja, kadang saya pun sulit untuk mengatasi data-race ini :)

rynzr replied:
Kalau sesuai contoh di atas, sepertinya itu masalahnya ada di bagian lock database-nya. Jika level akses read yang dipilih bisa read uncommitted data, maka operasi read akan melakukan dirty read. Tapi kalau yang dipilih level yang lebih menjaga data maka harusnya tidak terjadi seperti contoh tersebut.

Saya juga sekalian lagi cari sih untuk case bagaimana kemungkinan data race-nya akan terjadi. Walaupun di [dokumentasi library-nya](https://golang.org/pkg/database/sql/#DB ) tertulis:

> DB is a database handle representing a pool of zero or more underlying connections. It's safe for concurrent use by multiple goroutines.

dan di link ini https://notes.shichao.io/gopl/ch9/#chapter-9-concurrency-with-shared-variables :

> Indeed, concurrency-safe types are the exception rather than the rule, so you should access a variable concurrently only if the documentation for its type says that this is safe.

Masih ada kurang aman untuk concurrency ya? Kalau seperti itu berarti perlu ada mutex lagi atau copy value. Tapi dengan seperti itu pasti menambah overhead. Untuk copy value, saya harus coba dulu sih, ini nantinya jadi connection baru atau tidak, kalau terhitung connection baru kemungkinan overhead bertambah banyak karena akan open dan close connection terus.

reza-putra replied:
iya jika kamu pakai struct dari library `database/sql`. tapi kamu menghandle DB connection pooler sendiri pada library mu jadi kamu perlu meyakinkan bahwa ini safe for concurrent.

jika kamu lihat source code dari `database/sql` dia menggunakan sync.Mutex untuk mengatasi concurrent, lihat disini https://golang.org/src/database/sql/sql.go?s=9571:11341#L340.

iya kamu bisa tambahkan mutext dan lakukan test apakah overhead nya worhty atau tidak.

rynzr replied:
Bukannya kalau implementasi sekarang sama saja dengan menggunakan struct dari `database/sql`. Karena sebenarnya `ConnectionPooler` hanya contract untuk connection ke DB yang mana implementasinya, termasuk untuk concurrency, di-handle oleh apapun yang mengimplementasikan `ConnectionPooler`. Dalam kasus ini yang mengimplementasikan adalah struct dari `database/sql`, jadi bukannya sebenernya sama saja ya? Dan kalau ditambahkan mutex lagi jadi dua kali. Kalau ditambahkan juga jadinya malah lock seluruh resource, bukan inti resource untuk connection ke DB yang sudah di-handle oleh library `database/sql`. Ada yang salah ngga dari pemahaman saya seperti itu?

reza-putra replied:
owh kamu hanya bungkus database/sql nya yah. i see, saya pikir kamu handle connection pooler sendiri, line code ini https://github.com/nandaryanizar/fury/blob/feature/query/connection_pool.go#L32 sepertinya terlewat oleh saya. sip

reza-putra commented on db_test.go:
```diff
@@ -0,0 +1,350 @@
+					UserID: 5,
+				},
+			},
+			&[]*Account{
```
ini akan failed kan jika insert dengan berbeda type struct contoh:

```
// in top of the file
type Account2 struct {
	UserID    int `fury:"primary_key,auto_increment"`
	Username  string
	Password  string
	Email     string
	CreatedOn time.Time
	LastLogin time.Time
}
```

```
// lalu pada testcase nya di tambahkan 
&[]interface{}{
				&Account{
					UserID:    4,
					Username:  "test4",
					Password:  "test4",
					Email:     "test4@test.com",
					CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
					LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
				},
				&Account{
					UserID:    5,
					Username:  "test5",
					Password:  "test5",
					Email:     "test5@test.com",
					CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
					LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
				},
				&Account2{
					UserID:    5,
					Username:  "test5",
					Password:  "test5",
					Email:     "test5@test.com",
					CreatedOn: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
					LastLogin: time.Date(2016, 06, 22, 19, 10, 25, 0, time.FixedZone("", 0)),
				},
````

rynzr replied:
Ini karena design model-nya itu hanya untuk satu table. Jadi kalau dia terima selain pointer to struct atau pointer to slice dari pointer to struct akan error. Error message-nya dan comment-nya belum diganti, tulisannya masih `Error: expected pointer to struct, slice of pointer to struct or pointer to slice of pointer to struct, found`, harusnya cuma expect pointer to struct sama pointer to slice of pointer to struct aja

reza-putra replied:
sebaiknya jangan pakai type data `Interface{}` kosong untuk seperti ini, karena interface ini tidak miliki method dan akan menerima semua type data.  sebaiknya pakai `struct` atau inject nya ke setiap `Model`

contoh:

```
type Modeler interface{
  Insert()
  Update()
  Delete()
}
```

lalu implementasi nya
```
type Account struct {
	UserID    int `fury:"primary_key,auto_increment"`
         ................
}

func (a Account) Insert()
func (a Account) Delete()
func (a Account) Update()
```

rynzr replied:
Hmm.. kalau seperti ini bukannya nanti waktu dipakai ngga bisa jadi semacam generic gitu? karena method-nya masih di-define di struct masing-masing?

reza-putra replied:
iyes, karena go type language jadi deklartif seperti ini dan  go memang belum support generic. dengan seperti ini kita dapat handle use case khusus seperti jika user ingin menambahkan pekerjaan kita perlu mencek apakah user tsb sudah membatasi limit penambahaan pekerjaan atau jika satu entity ingin melakukan transaction sebelum insert ke db.

rynzr replied:
Bukannya kalau seperti ini kita violate SRP sama business logic-nya bisa kecampur sama repo ya?

reza-putra replied:
owh iya benar. sepertinya penjelasan saya diatas kurang benar. ini ada satu repo yang dapat kamu buat sebagai contoh dari maksud saya diatas. 

https://github.com/bxcodec/go-clean-arch/blob/master/author/repository.go

rynzr replied:
Kalau begitu terpisah sih, untuk design-nya kemungkinan ada dua. Kalau di repo itu kan dia tidak menggunakan library untuk CRUD, melainkan langsung query. Cara lain kita bisa define method standar di `Modeler` yang akan diimplementasikan di struct repo-nya. Berarti yang jadi parameter-nya itu struct repo-nya? Atau bagaimana bagusnya ya?

reza-putra replied:
iya yang menjadi paramater-nya itu struct repo nya.

rynzr replied:
Kalau misalkan seperti itu lebih baik untuk generate SQL bagaimana ya? Agak bingung untuk bagusnya di mana generator-nya ditaro dan apa saja kora-kira logic yang baik ditaro di setiap method `Insert`, dll.

reza-putra replied:
bisa sama saja kan seperti sekarang, tapi dari pada menggunakan empty interface pada  param insert model kamu dapat kamu dapat menggunakan interface dengan contract `insert`.

psudecode:
```
type ModelInserter interface{
func  Insert()
}

type ModelUpdater interface{
 func Update()
```

jika cara ini menurutmu kurang baik kamu dapat menggunakan cara lain yang lebih mudah untuk di implementasikan saat ini, karena approach lib kamu yang saat ini masih kurang baik, karena empty interface dapat menerima seluruh type data dan jika saya ingin menggunakan library ini, saya tidak bisa seacara intiutif bagaimana menggunakan library kamu tanpa membaca dokumentasi.

reza-putra commented on db_test.go:
```diff
@@ -0,0 +1,350 @@
+	}
+}
+
+func TestUpdateQuery(t *testing.T) {
```
jika seperti ini jika function ini dijalankan secara parallel dengan function `TestInsertQuery`  atau yang dijalankan hanya `TestUpdateQuery` maka akan error yah?

karena secara default command `go test` akan menjalankan testing secara parallel.
```
by default, it is set to the value of GOMAXPROCS.      Note that -parallel only applies within a single test binary.
     The 'go test' command may run tests for different packages  in parallel as well, according to the setting of the -p flag  (see 'go help build').
```

rule of thumb: setiap test harus ter-isolate & tidak tergantung dengan function lainnya

reza-putra replied:
jika ingin setiap test berjalan secara sequence. bisa ubah test nya menjadi seperti ini

```
func TestMainQuery(t *testing.T){
t.Run("insert", testInsertQuery)
t.Run("update", testUpdateQuery)
}
```
jangan lupa setiap functionnnya diubah menjadi private

rynzr replied:
Okeoke, nanti yang ini sekalian ditambahin

rynzr replied:
> rule of thumb: setiap test harus ter-isolate & tidak tergantung dengan function lainnya

Ini yang dimaksud context-nya kan? Maksudnya seperti jangan ada shared state variable. Kalau misalkan untuk test insert, update, delete mesti cek juga dengan find gpp kan ya?

Btw sudah ditambahin seeder-nya dan tiap method gunakan data yang beda kecuali yang find dan first karena hanya select. Kalau sudah begitu, ngga perlu dijadikan sequence kan?

reza-putra replied:
iyah bisa ditambahkan feeder untuk setiap test.

rynzr replied:
Oke kalau begitu

reza-putra commented on db_test.go:
```diff
@@ -0,0 +1,350 @@
+	}
+}
+
+func TestDeleteQuery(t *testing.T) {
```
ini juga sama yah seperti diatas.

reza-putra commented on db_test.go:
```diff
@@ -0,0 +1,350 @@
+		want interface{}
+	}{
+		{
+			&Account{},
```
ini maksud testCasenya apa yah? mungkin bisa ditambahkan comment.

rynzr replied:
Ini maksudnya cuma ngebandingin dua struct literal, kalau misalkan setelah di-delete waktu find lagi dapet return-nya atau ngga. Oke nanti ditambahin comment-nya

reza-putra replied:
owh diganti saja jadi `nil`

rynzr replied:
Ngga bisa kalau pakai nil, karena untuk test delete kan harus cek lagi ke database, jadi harus panggil `Find` lagi. Kalau nil dia akan error

reza-putra replied:
i see, oke kalo gitu

reza-putra commented on expression_test.go:
```diff
@@ -8,26 +8,25 @@ import (
 func TestExpressionToStringIfEmpty(t *testing.T) {
 	want := ""
-	have, _ := (&fury.Expression{}).ToString()
+	have, _, _ := (&fury.Expression{}).ToString()
```
kenapa deklarasinya struct nya tidak pakai new saja? jadi `new(fury.Expression)`

rynzr replied:
Ini bisa juga sih pakai new, karena kebiasaan lansung buat struct literal dan jarang pakai new jadi langsung pakai ini saja. Lebih baik diganti aja atau tetap begini?

reza-putra replied:
pakai `new` saja lebih simple dan readable

rynzr replied:
Ini sudah diganti yaa

reza-putra commented on expression_test.go:
```diff
@@ -8,26 +8,25 @@ import (
 
 	for _, tc := range cases {
-		_, err := fury.IsEqualsTo(tc.operand1, tc.operand2).ToString()
+		_, _, err := fury.IsEqualsTo(tc.operand1, tc.operand2).ToString()
```
ini `operand2` untuk apa yah? bukannya nanti akan diubah menjadi placeholder `"?"`  ?

rynzr replied:
Iya itu nanti diubah jadi placeholder, argument-nya juga disimpan di slice interface dan di-return. Dibuat seperti ini untuk check error kalau misalkan ada argument yang kurang, error message-nya masih belum terlalu informatif sekarang karena ngga dikasih tau error waktu buat expression dengan key dan operator apa, nanti bisa ditambahin. Sama dibuat seperti ini biar bisa reusable nanti expression-nya dan function `Where` juga bisa terima `Expression` dan `LogicalExpression`

reza-putra replied:
oke sip

reza-putra reviewed (changes requested):
kamu update README.md nya dulu yah, saya masih kurang paham beberapa penggunaannya.
file changes nya cukup banyak. reviewnya saya lanjut besok yah.

nandaryanizar commented:
> Pada file `README.md` tambahkan beberapa bagian ini.
> 
> 1. Description "deskripsi singkat tentang library ini"
> 2. Requirement ( dependency yang dibutuhkan e.g "harus intall go version 1.10 atau keatas )
> 3. How to use this library. "berikan contoh cara menggunakan library ini"

Oke nanti ditambahin readme-nya

nandaryanizar commented:
> module managment yang digunakan apa? jika `go mod` maka file `go.mod` dan `go.sum` nya juga di commit dan push.

Ini masih belum pake `go mod`, masih develop di `$GOPATH` biasa

evo3cx commented:
> > module managment yang digunakan apa? jika `go mod` maka file `go.mod` dan `go.sum` nya juga di commit dan push.
> 
> Ini masih belum pake `go mod`, masih develop di `$GOPATH` biasa

oke, kalo gitu tambahin `go mod` nya yah. bisa lihat documentnya di `go help mod`

rynzr committed:
Change after review

- Add README.md content
- Load config from environment variable
- Use go mod
- Change few codes after review

nandaryanizar commented:
> > > module managment yang digunakan apa? jika `go mod` maka file `go.mod` dan `go.sum` nya juga di commit dan push.
> > 
> > 
> > Ini masih belum pake `go mod`, masih develop di `$GOPATH` biasa
> 
> oke, kalo gitu tambahin `go mod` nya yah. bisa lihat documentnya di `go help mod`

Sipsip,  `go mod` dan `README.md`-nya sudah ditambahin ya

rynzr committed:
Add color to README.yaml

Add go and yaml notation for color

rynzr committed:
Change README and add ConnectMock function

- Add comment to README.md at insert segment
- Add ConnectMock for mocking
